src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                  [this, &octx, shard, sf = std::move(sf)]() mutable {
+                      return log_and_ignore_exceptions(start_shard_fetch_worker(
+                        shard,
+                        octx,
+                        std::move(sf),
+                        // Require that a worker returns more data than before.
+                        // Otherwise it'll return right away with the previous
+                        // result.
+                        _last_result_size[shard] + 1));
+                  });
+            }
+        }
+    }
+
+private:
+    /**
+     * Determines if the should_stop_fetch() condition should be checked again.
+     * The return future is set if;
+     * - octx.deadline has been reached.
+     * - octx.rctx.abort_source has been aborted.
+     * - one of the shard workers has returned results.
+     */
+    ss::future<> progress_conditions(op_context& octx) {
+        ss::promise<> timeout_pr, rctx_abort_pr;
+
+        auto abort_sub_opt = octx.rctx.abort_source().subscribe(
+          [&rctx_abort_pr]() noexcept { rctx_abort_pr.set_value(); });
+
+        if (!abort_sub_opt) {
+            rctx_abort_pr.set_value();
+        }
+
+        ss::timer<model::timeout_clock> s{
+          [&timeout_pr] { timeout_pr.set_value(); }};
+
+        if (octx.deadline) {
+            s.arm(octx.deadline.value());
+        }
+
+        co_await ss::when_any(
+          ignore_exceptions(_has_data.wait()),
+          ignore_exceptions(timeout_pr.get_future()),
+          ignore_exceptions(rctx_abort_pr.get_future()));
+    }
+
+    /*
+     * `start_shard_fetch_worker` executes on the coordinator shard. It builds
+     * the `shard_local_fetch_context` struct needed to start the
+     * shard_fetch_worker then makes a cross shard call to the shard fetch
+     * worker if needed.
+     *
+     * It then waits until the shard_fetch_worker completes
+     * and then notifies the fetch coordinator.
+     */
+    ss::future<> start_shard_fetch_worker(
+      ss::shard_id shard,
+      op_context& octx,
+      shard_fetch fetch,
+      size_t min_fetch_bytes) {
+        // if over budget skip the fetch.
+        if (octx.bytes_left <= 0) {
+            co_return;
+        }
+        // no requests for this shard, do nothing
+        if (fetch.empty()) {
+            co_return;
+        }
+
+        const bool foreign_read = shard != ss::this_shard_id();
+
+        auto results = co_await octx.rctx.partition_manager().invoke_on(
+          shard,
+          octx.ssg,
+          [this,
+           min_fetch_bytes,
+           foreign_read,
+           configs = fetch.requests,
+           &octx](cluster::partition_manager& mgr) mutable {
+              // &octx is captured only to immediately use its accessors here
+              // so that there is a list of all objects accessed next to
+              // `invoke_on`. This is meant to help avoiding unintended cross
+              // shard access
```

Comment:
It's a stale comment sorry. Technically references to `octx` and `this` are briefly captured and used on a different shard to pull out data local to that other shard. It should be safe though since none of the data being read is modified by either shard and both are protected from deletion by a gate that can't be closed until the cross-shard call returns.